packages/common/infra/src/sync/indexer/README.md

```diff
@@ -0,0 +1,147 @@
+# index
+
+Search engine abstraction layer for AFFiNE.
+
+## Using
+
+1. Define schema
+
+First, we need to define the shape of the data. Currently, there are the following data types.
+
+- 'Integer'
+- 'Boolean'
+- 'FullText': for full-text search, it will be tokenized and stemmed.
+- 'String': for exact match search, e.g. tags, ids.
+
+```typescript
+const schema = {
+  title: 'FullText',
+  tag: 'String',
+  size: 'Integer',
+} as Schema;
+```
+
+> **Array type**
+> All types can contain one or more values, so each field can store an array.
+
+2. Pick a backend
+
+Currently, there are two backends available.
+
+- `MemoryIndex`: in-memory indexer, useful for testing.
+- `IndexedDBIndex`: persistent indexer using IndexedDB.
+
+> **Underlying Data Table**
+> Some back-end processes need to maintain underlying data tables, including table creation and migration. This operation should be silently executed the first time the indexer is invoked.
+> Callers do not need to worry about these details.
+>
+> This design conforms to the usual conventions of search engine APIs, such as in Elasticsearch: https://www.elastic.co/guide/en/elasticsearch/reference/current/array.html
+
+3. Write data
+
+Write data to the indexer. you need to start a write transaction by `await index.write()` first and then complete the batch write through `await writer.commit()`.
+
+> **Transactional**
+> Typically, the indexer does not provide transactional guarantees; reliable locking logic needs to be implemented at a higher level.
+
+```typescript
+const indexer = new IndexedDBIndex(schema);
+
+const writer = await index.write();
+writer.insert(
+  Document.from('id', {
+    title: 'hello world',
+    tag: ['doc', 'page'],
+    size: '100',
+  })
+);
+await writer.commit();
+```
+
+4. Search data
+
+To search for content in the indexer, you need to use a specific **query language**. Here are some examples:
+
+```typescript
+// match title == 'hello world'
+{
+  type: 'match',
+  field: 'title',
+  match: 'hello world',
+}
+
+// match title == 'hello world' && tag == 'doc'
+{
+    type: 'boolean',
+    occur: 'must',
+    queries: [
+        {
+            type: 'match',
+            field: 'title',
+            match: 'hello world',
+        },
+        {
+            type: 'match',
+            field: 'tag',
+            match: 'doc',
+        },
```

Comment:
Does the `FullText` type support both exact match and fuzzy search? 

It seems that the syntax for fuzzy search and exact match is exactly the same. Is it the case that for `FullText` it is tokenized search, and for `String` it is exact match?

Reply:
fulltext always fuzzy search, string always exact search, and there is currently no way to define whether you need fuzzy or exact search.

They all use `match`, and the final behavior is defined by the internal implementation.

This is due to the limitations of the underlying implementation. For example, `elasticsearch` cannot choose which search method to use at runtime, and must decide before index.